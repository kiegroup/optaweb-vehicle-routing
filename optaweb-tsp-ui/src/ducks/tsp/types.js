/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const SOLUTION_UPDATES_DATA = 'SOLUTION_UPDATES_DATA';
const DELETE_LOCATION = 'DELETE_LOCATION';
const ADD_LOCATION = 'ADD_LOCATION';
const ADD_DEMO_LOCATION = 'ADD_DEMO_LOCATION';
const WS_CONNECT = 'WS_CONNECT';
const WS_CONNECT_SUCCESS = 'WS_CONNECT_SUCCESS';
const WS_CONNECT_FAILURE = 'WS_CONNECT_FAILURE';


export default {
  SOLUTION_UPDATES_DATA,
  DELETE_LOCATION,
  ADD_LOCATION,
  ADD_DEMO_LOCATION,
  WS_CONNECT,
  WS_CONNECT_SUCCESS,
  WS_CONNECT_FAILURE,
};
