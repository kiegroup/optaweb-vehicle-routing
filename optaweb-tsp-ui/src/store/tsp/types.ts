/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const SOLUTION_UPDATES_DATA = 'SOLUTION_UPDATES_DATA';
export const DELETE_LOCATION = 'DELETE_LOCATION';
export const ADD_LOCATION = 'ADD_LOCATION';
export const ADD_DEMO_LOCATION = 'ADD_DEMO_LOCATION';
export const WS_CONNECT = 'WS_CONNECT';
export const WS_CONNECT_SUCCESS = 'WS_CONNECT_SUCCESS';
export const WS_CONNECT_FAILURE = 'WS_CONNECT_FAILURE';

export type SOLUTION_UPDATES_DATA = typeof SOLUTION_UPDATES_DATA;
export type DELETE_LOCATION = typeof DELETE_LOCATION;
export type ADD_LOCATION = typeof ADD_LOCATION;
export type ADD_DEMO_LOCATION = typeof ADD_DEMO_LOCATION;
export type WS_CONNECT = typeof WS_CONNECT;
export type WS_CONNECT_SUCCESS = typeof WS_CONNECT_SUCCESS;
export type WS_CONNECT_FAILURE = typeof WS_CONNECT_FAILURE;

export enum WS_CONNECTION_STATE {
  OPEN,
  CLOSED,
  ERROR,
}
export interface ILatLng {
  lat: number;
  lng: number;
}

export interface IdLatLng extends ILatLng {
  id: number;
}
export interface ITSPRoute {
  route: IdLatLng[];
  domicileId: number;
  distance?: string;
}

export interface IWSConnection{
  ws: WS_CONNECTION_STATE;
}
