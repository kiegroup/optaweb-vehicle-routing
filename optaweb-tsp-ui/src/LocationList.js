import PropTypes from 'prop-types';
import React from 'react';
import Location from './Location';

function LocationList({ route, domicileId, removeHandler, selectHandler }) {
  return (
    <div className={'leaflet-top leaflet-left leaflet-touch'}>
      <div className={'leaflet-control leaflet-bar w5 bg-white '}>
        {
          route.length === 0 ?
            <div className={'tc ma2'}>Click map to add locations</div> :
            route
              .slice(0) // clone the array
              .sort((a, b) => a.id - b.id) // because sort is done in place (that would affect the route)
              .map(location => (
                <Location
                  key={location.id}
                  id={location.id}
                  removeDisabled={route.length > 1 && location.id === domicileId}
                  removeHandler={removeHandler}
                  selectHandler={selectHandler}
                />
              ))
        }
      </div>
    </div>
  );
}

LocationList.propTypes = {
  route: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    lat: PropTypes.number.isRequired,
    lng: PropTypes.number.isRequired,
  })).isRequired,
  domicileId: PropTypes.number.isRequired,
  removeHandler: PropTypes.func.isRequired,
  selectHandler: PropTypes.func.isRequired,
};

export default LocationList;
