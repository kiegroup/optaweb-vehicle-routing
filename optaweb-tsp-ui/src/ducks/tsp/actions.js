

import types from './types';

function makeActionCreator(type, ...argNames) {
  return function actionCreator(...args) {
    const action = { type };
    argNames.forEach((arg, index) => {
      action[argNames[index]] = args[index];
    });
    return action;
  };
}


// consider https://github.com/infinitered/reduxsauce  for DRY approach

const actions = {
  addLocation: makeActionCreator(types.ADD_LOCATION, 'value'),

  deleteLocation: makeActionCreator(types.DELETE_LOCATION, 'value'),

  updateTPSSolution: makeActionCreator(types.SOLUTION_UPDATES_DATA, 'solution'),


  initWsConnection: makeActionCreator(types.WS_CONNECT, 'value'),
  wsConnectionSuccess: makeActionCreator(types.WS_CONNECT_SUCCESS, 'value'),
  wsConnectionFailure: makeActionCreator(types.WS_CONNECT_FAILURE, 'value'),
};


export default actions;
