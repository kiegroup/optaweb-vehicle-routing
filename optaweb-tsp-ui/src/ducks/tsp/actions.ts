/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { GPSLocation, ADD_LOCATION, DELETE_LOCATION, SOLUTION_UPDATES_DATA, WS_CONNECT, WS_CONNECT_SUCCESS, WS_CONNECT_FAILURE, TSPStore } from "./types";
import { Client, Frame } from "webstomp-client";

// Dropped action creator maker due to more complex type handling, a strategy will be defined once needed

export type AddLocationAction = {
  type: typeof ADD_LOCATION;
  value: GPSLocation;
};

export type DeleteLocationAction = {
  type: typeof DELETE_LOCATION;
  value: number;
};

export type UpdateTSPSolutionAction = {
  type: typeof SOLUTION_UPDATES_DATA;
  solution: TSPStore;
};

export type InitWsConnectionAction = {
  type: typeof WS_CONNECT;
  value: string;
};

export type WsConnectionSuccessAction = {
  type: typeof WS_CONNECT_SUCCESS;
  value: Client;
};

export type WsConnectionFailureAction = {
  type: typeof WS_CONNECT_FAILURE;
  value: Frame | CloseEvent;
};

const addLocation = (location: GPSLocation): AddLocationAction => ({
  type: ADD_LOCATION,
  value: location
});

const deleteLocation = (id: number): DeleteLocationAction => ({
  type: DELETE_LOCATION,
  value: id
});

const updateTSPSolution = (
  solution: TSPStore
): UpdateTSPSolutionAction => ({
  type: SOLUTION_UPDATES_DATA,
  solution
});

const initWsConnection = (socketUrl: string): InitWsConnectionAction => ({
  type: WS_CONNECT,
  value: socketUrl
});

const wsConnectionSuccess = (
  webstompSocket: Client
): WsConnectionSuccessAction => ({
  type: WS_CONNECT_SUCCESS,
  value: webstompSocket
});

const wsConnectionFailure = (
  err: Frame | CloseEvent
): WsConnectionFailureAction => ({
  type: WS_CONNECT_FAILURE,
  value: err
});

export default {
  addLocation,
  deleteLocation,
  updateTSPSolution,
  initWsConnection,
  wsConnectionSuccess,
  wsConnectionFailure
};
