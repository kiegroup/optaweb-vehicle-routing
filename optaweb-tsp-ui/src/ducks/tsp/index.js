import tspReducer from './reducers';

export { default as tspOperations } from './operations';
export { default as tspTypes } from './types';
export default tspReducer;
