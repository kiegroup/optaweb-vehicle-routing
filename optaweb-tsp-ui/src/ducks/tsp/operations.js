/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import SockJS from 'sockjs-client';
import webstomp from 'webstomp-client';
import Creators from './actions';

const {
  addLocation,
  deleteLocation,
  updateTPSSolution,
  initWsConnection,
  wsConnectionSuccess,
  wsConnectionFailure }
  = Creators;


let webSocket;
let stompClient;

function mapEventToActions(dispatch) {
  stompClient.subscribe('/topic/route', (message) => {
    const tsp = JSON.parse(message.body);
    dispatch(updateTPSSolution(tsp));
  });
}

function connectWs(store, socketUrl) {
  return (dispatch) => {
    webSocket = new SockJS(socketUrl);
    stompClient = webstomp.over(webSocket, { debug: true });

    dispatch(initWsConnection(socketUrl));
    stompClient.connect(
      {}, // no headers
      () => { // on connection, subscribe to the route topic
        dispatch(wsConnectionSuccess(stompClient));
        mapEventToActions(store.dispatch);
      }, (err) => { // on error, schedule a reconnection attempt
        dispatch(wsConnectionFailure(err));
        setTimeout(() => connectWs(store, socketUrl), 1000);
      });
  };
}


const addLocationOp = (location, demo = false) => (dispatch) => {
  if (demo) {
    stompClient.send('/app/demo');
    dispatch(addLocation(location));
    return;
  }
  if (!location) {
    return;
  }

  stompClient.send('/app/place', JSON.stringify(location));
  dispatch(addLocation(location));
};

const deleteLocationOp = locationId => (dispatch) => {
  if (!locationId) {
    return;
  }
  stompClient.send(`/app/place/${locationId}/delete`, JSON.stringify(locationId));
  dispatch(deleteLocation(locationId));
};

export default {
  connect: connectWs,
  addLocation: addLocationOp,
  deleteLocation: deleteLocationOp,

};

