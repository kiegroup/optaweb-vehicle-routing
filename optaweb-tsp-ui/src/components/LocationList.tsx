/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Card, CardBody, CardHeader } from '@patternfly/react-core';
import * as React from 'react';
import { ITSPRoute } from '../store/tsp/types';
import Location from './Location';
import TripData from './TripData';

export interface ILocationListProps extends ITSPRoute {
  removeHandler: (id: number) => void;
  selectHandler: (e: any) => void;
  loadHandler: () => void;
  maxDistance: number;
}

const renderEmptyLocationList = ({ loadHandler }: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>Click map to add locations</CardHeader>
      <CardBody>
        <Button type="button" style={{ width: '100%' }} onClick={loadHandler}>
          Load 40 European cities
        </Button>
      </CardBody>
    </Card>
  );
};

const renderLocationList = ({
  route,
  domicileId,
  distance = '',
  removeHandler,
  selectHandler,
  loadHandler,
  maxDistance,
}: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>
        Distance: {distance}
        <br />
        Locations: {route.length}
        <TripData
          maxDistance={maxDistance}
          distance={parseInt(distance, 10) || maxDistance}
        />
        <hr />
      </CardHeader>
      <CardBody>
        {/*
               The calculated maxHeight is a hack because the constant 116px depends
               on the height of Distance and Locations rows (above) and individual location rows.
               */}
        <div style={{ maxHeight: 'calc(100vh - 195px)', overflowY: 'auto' }}>
          {route
            .slice(0) // clone the array because
            // sort is done in place (that would affect the route)
            .sort((a, b) => a.id - b.id)
            .map(location => (
              <Location
                key={location.id}
                id={location.id}
                removeDisabled={route.length > 1 && location.id === domicileId}
                removeHandler={removeHandler}
                selectHandler={selectHandler}
              />
            ))}
        </div>
      </CardBody>
    </Card>
  );
};

const LocationList: React.SFC<ILocationListProps> = (
  props: ILocationListProps
) => {
  return (
    <div
      className="leaflet-top leaflet-left leaflet-touch"
      style={{ zIndex: 500 }}
    >
      <div className="leaflet-control leaflet-bar">
        {props.route.length === 0
          ? renderEmptyLocationList(props)
          : renderLocationList(props)}
      </div>
    </div>
  );
};

export default LocationList;
