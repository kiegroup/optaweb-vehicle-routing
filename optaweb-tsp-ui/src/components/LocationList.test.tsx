/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import * as types from '../store/tsp/types';
import LocationList from './LocationList';
describe('Location List Component', () => {
  it('should render correctly with no routes', () => {
    const props = {
      distance: '10',
      domicileId: 1,
      loadHandler: jest.fn(),
      maxDistance: 1,
      removeHandler: jest.fn(),
      route: [],
      selectHandler: jest.fn(),
      ws: types.WS_CONNECTION_STATE.OPEN
  
    };
    expect.assertions(2);
    const locationList = shallow(<LocationList {...props} />);
    expect(locationList).toMatchSnapshot();

    locationList.find('Button').simulate('click');

    expect(props.loadHandler).toHaveBeenCalledTimes(1);
  });

  it('should render correctly with a few routes', () => {
    const props = {
      distance: '10',
      domicileId: 1,
      loadHandler: jest.fn(),
      maxDistance: 1,
      removeHandler: jest.fn(),
      route: [
        {
          id: 1,
          lat: 1.345678,
          lng: 1.345678,
        },
        {
          id: 2,
          lat: 2.345678,
          lng: 2.345678,
        },
        {
          id: 3,
          lat: 3.676111,
          lng: 3.568333,
        },
      ],
      selectHandler: jest.fn(),
      ws: types.WS_CONNECTION_STATE.OPEN
    };
    expect.assertions(2);
    const locationList = shallow(<LocationList {...props} />);
    expect(locationList).toMatchSnapshot();

    expect(locationList.find('Location')).toHaveLength(props.route.length);
  });
});
