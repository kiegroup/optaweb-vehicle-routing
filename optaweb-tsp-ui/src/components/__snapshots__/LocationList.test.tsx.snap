// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location List Component should render correctly with a few routes 1`] = `
<div
  className="leaflet-top leaflet-left leaflet-touch"
  style={
    Object {
      "zIndex": 500,
    }
  }
>
  <div
    className="leaflet-control leaflet-bar"
  >
    <Card
      className=""
      component="article"
    >
      <CardHeader
        className=""
        component="div"
      >
        Distance: 
        10
        <br />
        Locations: 
        3
        <TripData
          distance={10}
          maxDistance={1}
        />
        <br />
        <Button
          aria-label={null}
          className=""
          component="button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[MockFunction]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="button"
          variant="primary"
        >
          Clear
        </Button>
        <br />
      </CardHeader>
      <CardBody
        className=""
        component="div"
      >
        <div
          style={
            Object {
              "maxHeight": "calc(100vh - 24px - 24px - 8px - 196px)",
              "overflowY": "auto",
            }
          }
        >
          <Location
            id={1}
            key="1"
            removeDisabled={true}
            removeHandler={[MockFunction]}
            selectHandler={[MockFunction]}
          />
          <Location
            id={2}
            key="2"
            removeDisabled={false}
            removeHandler={[MockFunction]}
            selectHandler={[MockFunction]}
          />
          <Location
            id={3}
            key="3"
            removeDisabled={false}
            removeHandler={[MockFunction]}
            selectHandler={[MockFunction]}
          />
        </div>
      </CardBody>
    </Card>
  </div>
</div>
`;

exports[`Location List Component should render correctly with no routes 1`] = `
<div
  className="leaflet-top leaflet-left leaflet-touch"
  style={
    Object {
      "zIndex": 500,
    }
  }
>
  <div
    className="leaflet-control leaflet-bar"
  >
    <Card
      className=""
      component="article"
    >
      <CardHeader
        className=""
        component="div"
      >
        Click map to add locations
      </CardHeader>
      <CardBody
        className=""
        component="div"
      >
        <Button
          aria-label={null}
          className=""
          component="button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[MockFunction]}
          style={
            Object {
              "width": "100%",
            }
          }
          type="button"
          variant="primary"
        >
          Load demo
        </Button>
      </CardBody>
    </Card>
  </div>
</div>
`;
